### PR TITLE
PrimeRust: PrimeSieve Rust implementation.

### PR DESCRIPTION
**Pull request by Azgrom**
PrimeSieve Rust implementation.

Based this code on PrimeSievePy, PrimeCPP and on #107 @joshallen64's Rust pull request.

Just got the following results:
Passes: 11630, Time: 10.000201941s, Avg: 859.862µs, Limit: 1000000, Count1: 78498, Count2: 78498, Valid: true

My specs are: i7 9750H, 32GB RAM, Manjaro KDE

Project created and managed with [Cargo](https://www.rust-lang.org/tools/install)

Test was run with: `$ cargo run --release`
Code formatted with: `$ cargo fmt`

PS: _There is a little bug with the counts at 1000 limit. When run at 1_000_000 limit the counting goes perfect._

### CODE REVIEW

**Comment by marghidanu** [comment]
For some reason, this has a conflict on a Java file. Please review this locally and look at the CONTRIBUTING.md document for some of the contribution guidelines.

**Comment by mike-barber** [comment]
I think this might be due to this branch having been based on `main` originally and has all those changes in it as well, which I think we're _not_ ready to merge yet, but it's in the works.

To fix this here, I think this branch will need to be rebased off `drag-race`, and few other changes
- directory should be `PrimeRust/solution_2`
- PR should not alter anything external to your target directory
- go through CONTRIBUTING.md: there are some notes there about
  - the README
  - output format (for automatic benchmarking)

If it helps, have a look at my existing Rust implementation in `PrimeRust/solution_1` -- and feel free to re-use the Dockerfile; we'll be using this ultimately to run automated benchmarks and collect results.

**Comment by Azgrom** [comment]
> For some reason, this has a conflict on a Java file. Please review this locally and look at the CONTRIBUTING.md document for some of the contribution guidelines.

Yeah, it was as @mike-barber stated. I branched my solution from `main` HEAD, and did not payed attention that work was being done at `drag-race` branch. I am sorry for this mistake.

Just done what @mike-barber suggested. This solution is not as polished as his, so I will still have to improve it.

**Comment by mike-barber** [comment]
What I like about your solution is that it's a good straightforward Rust solution, and a really good example for new people interested in Rust to look at. Mine is doing quite a lot of things, so it's a bit complicated in comparison.

**Comment by Azgrom** [comment]
> What I like about your solution is that it's a good straightforward Rust solution, and a really good example for new people interested in Rust to look at. Mine is doing quite a lot of things, so it's a bit complicated in comparison.

Still, this solution performs 1/6 of yours, since your test runs at 5s.
With the same initial specification:
> Just got the following results:
Passes: 11630, Time: 10.000201941s, Avg: 859.862µs, Limit: 1000000, Count1: 78498, Count2: 78498, Valid: true
> 
> My specs are: i7 9750H, 32GB RAM, Manjaro KDE

Those were the results:
```bash
$ ./run-docker.sh 

Computing primes to 1000000 on 1 thread for 5 seconds.
byte-storage;7579;5.0009570122;1
byte-storage    Passes: 7579, Threads: 1, Time: 5.0009570122, Average: 0.0006598439, Limit: 1000000, Counts: 78498, Valid: Pass

Computing primes to 1000000 on 1 thread for 5 seconds.
bit-storage;6616;5.0007867813;1
bit-storage     Passes: 6616, Threads: 1, Time: 5.0007867813, Average: 0.0007558626, Limit: 1000000, Counts: 78498, Valid: Pass

Computing primes to 1000000 on 12 threads for 5 seconds.
byte-storage    Passes: 26610, Threads: 12, Time: 5.0017600060, Average: 0.0001879654, Limit: 1000000, Counts: 78498, Valid: Pass

Computing primes to 1000000 on 12 threads for 5 seconds.
byte-storage;26610;5.0017600060;12
bit-storage;30731;5.0019030571;12
bit-storage     Passes: 30731, Threads: 12, Time: 5.0019030571, Average: 0.0001627641, Limit: 1000000, Counts: 78498, Valid: Pass
```

Wonder where would be the bottleneck on `solution_2` 🤔

**Comment by mike-barber** [comment]
Hi @Azgrom 

I can't see much of a performance difference on my hardware; your code is only around 20% slower. I'll add some comments to the PR inline for specific suggestions as there are a couple of things you could tweak :) 

My code: 
```
byte-storage    Passes: 18194, Threads: 1, Time: 5.0000767708, Average: 0.0002748201, Limit: 1000000, Counts: 78498, Valid: Pass
```

Your code, adjusted to 5 seconds (as required in `CONTRIBUTING.md` -- keep an eye on the output format requirements too)
```
Passes: 14103, Time: 5.0001065s, Avg: 354.541µs, Limit: 1000000, Count1: 78498, Count2: 78498, Valid: true
```

Of course you need to compare the right results: your solution is single-threaded only and uses `Vec<bool>` for storage. So the corresponding one to check in my results is the `byte storage` result, with `1 thread`. If you compare it with the multithreaded solution, I guess you'd expect about 1/6 of the performance because you've got 6 cores in your processor. Essentially, compare your results with the first line of my results.

File: PrimeRust/solution_2/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+[package]
+name = "prime_rust"
+version = "0.1.0"
+authors = ["Rafael Lúcio <rafaellucio@pm.me>"]
+edition = "2018"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
```
Comment (mike-barber) [inline PrimeRust/solution_2/Cargo.toml]: I'd suggest enabling link-time optimisation here:
```
[profile.release]
opt-level = 3
lto = true
codegen-units = 1
```

File: PrimeRust/solution_2/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    let mut passes = 0;
+    let time_start = Instant::now();
+
+    loop {
+        let mut prime = PrimeSieve::new(1_000_000);
+        prime.run_sieve();
+
+        passes += 1;
+
+        if (Instant::now() - time_start) >= Duration::new(10, 0) {
```
Comment (mike-barber) [inline PrimeRust/solution_2/src/main.rs]: Aim for 5 seconds -- easier to compare with all the other solutions.

File: PrimeRust/solution_2/src/prime_object.rs [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+            let mut num = factor;
+            while num < q {
+                if self.raw_bits[num] {
+                    factor = num;
+                    break;
+                }
+                num += 2;
+            }
+
+            num = factor * 3;
```
Comment (mike-barber) [inline PrimeRust/solution_2/src/prime_object.rs]: You can start from `factor * factor` here. @Guiorgy spotted this on my original implementation too -- and it's been updated in the C++ version on the `main` branch.

**Review by mike-barber — COMMENTED** [review]
Some suggestions to improve performance

**Comment by Azgrom** [comment]
> Some suggestions to improve performance

Thank you! Just added each tip in the above commits. The combination of those changes increased around 500 passes at the limit of 5s.

**Comment by marghidanu** [comment]
The numbers look great! Please add a Dockerfile so we can include it in the report.

**Comment by marghidanu** [comment]
This looks good to me, also it builds really fast. I think it's ready to merge.

**Pull request merged**